lib/bitcoin.py

```diff
@@ -754,7 +780,299 @@ def decrypt_message(self, encrypted):
+        try:
+            ECPubkey(b)
+            return True
+        except:
+            return False
+
+def generator():
+    return ECPubkey.from_point(generator_secp256k1)
+
+
+def point_at_infinity():
+    return ECPubkey(None)
+
+
+def sig_string_from_der_sig(der_sig: bytes, order=CURVE_ORDER) -> bytes:
+    r, s = ecdsa.util.sigdecode_der(der_sig, order)
+    return ecdsa.util.sigencode_string(r, s, order)
+
+
+def der_sig_from_sig_string(sig_string: bytes, order=CURVE_ORDER) -> bytes:
+    r, s = ecdsa.util.sigdecode_string(sig_string, order)
+    return ecdsa.util.sigencode_der_canonize(r, s, order)
+
+
+def der_sig_from_r_and_s(r: int, s: int, order=CURVE_ORDER) -> bytes:
+    return ecdsa.util.sigencode_der_canonize(r, s, order)
+
+
+def get_r_and_s_from_der_sig(der_sig: bytes, order=CURVE_ORDER) -> Tuple[int, int]:
+    r, s = ecdsa.util.sigdecode_der(der_sig, order)
+    return r, s
+
+
+def get_r_and_s_from_sig_string(sig_string: bytes, order=CURVE_ORDER) -> Tuple[int, int]:
+    r, s = ecdsa.util.sigdecode_string(sig_string, order)
+    return r, s
+
+
+def sig_string_from_r_and_s(r: int, s: int, order=CURVE_ORDER) -> bytes:
+    return ecdsa.util.sigencode_string_canonize(r, s, order)
+
+
+def is_secret_within_curve_range(secret: Union[int, bytes]) -> bool:
+    if isinstance(secret, bytes):
+        secret = string_to_number(secret)
+    return 0 < secret < CURVE_ORDER
+
+
+def construct_sig65(sig_string: bytes, recid: int, is_compressed: bool) -> bytes:
+    comp = 4 if is_compressed else 0
+    return bytes([27 + recid + comp]) + sig_string
+
+
+class ECPrivkey(ECPubkey):
+
+    def __init__(self, privkey_bytes: bytes):
+        assert_bytes(privkey_bytes)
+        if len(privkey_bytes) != 32:
+            raise Exception('unexpected size for secret. should be 32 bytes, not {}'.format(len(privkey_bytes)))
+        secret = string_to_number(privkey_bytes)
+        if not is_secret_within_curve_range(secret):
+            raise InvalidECPointException('Invalid secret scalar (not within curve order)')
+        self.secret_scalar = secret
+
+        point = generator_secp256k1 * secret
+        super().__init__(point_to_ser(point))
+        self._privkey = ecdsa.ecdsa.Private_key(self._pubkey, secret)
+
+    @classmethod
+    def from_secret_scalar(cls, secret_scalar: int):
+        secret_bytes = number_to_string(secret_scalar, CURVE_ORDER)
+        return ECPrivkey(secret_bytes)
+
+    @classmethod
+    def from_arbitrary_size_secret(cls, privkey_bytes: bytes):
+        """This method is only for legacy reasons. Do not introduce new code that uses it.
+        Unlike the default constructor, this method does not require len(privkey_bytes) == 32,
+        and the secret does not need to be within the curve order either.
+        """
+        return ECPrivkey(cls.normalize_secret_bytes(privkey_bytes))
+
+    @classmethod
+    def normalize_secret_bytes(cls, privkey_bytes: bytes) -> bytes:
```

Comment:
(this also is unused aside from the deprecated function)

Reply:
I just landed in NY last night— after I get my bearings here I’ll clean this up.

Thanks for the comment.